### PR TITLE
fix: fix password form validation and update password logic for oauth…

### PR DESCRIPTION
**Pull request by luckyklyist**
Issue no : #620 

Fixed up the password form validation and updated the password logic for OAuth in response to issue #620. Introduced isPasswordNull routes and improved the password update process by checking for null values in the password field.

https://github.com/documenso/documenso/assets/35479077/b0da4339-6d80-4d2b-b0d2-24edbdba0ee5

### CODE REVIEW

File: packages/lib/server-only/user/update-password.ts [hunk 1]
```diff
@@ -3,50 +3,71 @@ import { compare, hash } from 'bcrypt';
 import { prisma } from '@documenso/prisma';
 
 import { SALT_ROUNDS } from '../../constants/auth';
+import { isPasswordNull } from './is-password-null';
 
 export type UpdatePasswordOptions = {
   userId: number;
   password: string;
-  currentPassword: string;
+  currentPassword?: string;
```
Comment (Mythie) [inline packages/lib/server-only/user/update-password.ts:11]: `currentPassword?: string | null;` right?

File: packages/lib/server-only/user/update-password.ts [hunk 2]
```diff
@@ -3,50 +3,71 @@ import { compare, hash } from 'bcrypt';
...
 };
 
 export const updatePassword = async ({
   userId,
   password,
   currentPassword,
 }: UpdatePasswordOptions) => {
+  console.log('userId', userId, 'update passowrd');
```
Comment (Mythie) [inline packages/lib/server-only/user/update-password.ts:19]: 🤓

File: packages/lib/server-only/user/update-password.ts [hunk 3]
```diff
@@ -3,50 +3,71 @@ import { compare, hash } from 'bcrypt';
...
   const user = await prisma.user.findFirstOrThrow({
     where: {
       id: userId,
     },
   });
 
-  if (!user.password) {
+  if (!user) {
     throw new Error('User has no password');
   }
```
Comment (Mythie) [inline packages/lib/server-only/user/update-password.ts:29]: This is probably redundant now

File: packages/lib/server-only/user/update-password.ts [hunk 4]
```diff
@@ -3,50 +3,71 @@ import { compare, hash } from 'bcrypt';
...
 
-  const isCurrentPasswordValid = await compare(currentPassword, user.password);
-  if (!isCurrentPasswordValid) {
-    throw new Error('Current password is incorrect.');
-  }
+  const isUserPasswordNull = await isPasswordNull({ id: userId });
```
Comment (Mythie) [inline packages/lib/server-only/user/update-password.ts:31]: This would mean we're retrieving the user twice, there should be no need for that?

File: packages/lib/server-only/user/update-password.ts [hunk 5]
```diff
@@ -3,50 +3,71 @@ import { compare, hash } from 'bcrypt';
...
+      const hashedNewPassword = await hash(password, SALT_ROUNDS);
 
-  return updatedUser;
+      const updatedUser = await prisma.user.update({
+        where: {
+          id: userId,
+        },
+        data: {
+          password: hashedNewPassword,
+        },
```
Comment (Mythie) [inline packages/lib/server-only/user/update-password.ts:67]: This should be simplified to only need an `if` branch if the password is not null where we can check and throw otherwise we proceed with the update. 

Right now it's doing the work twice and is a bit more confusing since `else` branches become hard to follow.

File: packages/trpc/server/profile-router/router.ts [hunk 6]
```diff
@@ -110,4 +111,15 @@ export const profileRouter = router({
...
+    try {
+      const isUserPasswordNull = await isPasswordNull({ id: ctx.user?.id as number });
+      return isUserPasswordNull;
+    } catch (err) {
+      throw new TRPCError({
+        code: 'BAD_REQUEST',
+        message: 'An error occurred while checking the user existence. Please try again.',
+      });
+    }
+  }),
```
Comment (Mythie) [inline packages/trpc/server/profile-router/router.ts:124]: We can probably just retrieve this during the server render and eliminate the need for the tRPC endpoint.

That said I do appreciate that you're just using the ctx here for the userId and not introducing it as a param which could be abused, good work!

File: apps/web/src/components/forms/password.tsx [hunk 7]
```diff
@@ -21,9 +21,11 @@ import { FormErrorMessage } from '../form/form-error-message';
...
   .object({
     currentPassword: z
-      .string()
-      .min(6, { message: 'Password should contain at least 6 characters' })
-      .max(72, { message: 'Password should not contain more than 72 characters' }),
+      .union([
+        z.string().min(6, { message: 'Password should contain at least 6 characters' }),
+        z.string().max(72, { message: 'Password should not contain more than 72 characters' }),
+      ])
+      .optional(),
```
Comment (Mythie) [inline apps/web/src/components/forms/password.tsx:28]: This still feels clunky, would it work as a nullable like you've done at the bottom of this PR?

**Review by Mythie — COMMENTED** [review]
If we go one level up from this form we should be able to retrieve the user there and pass in a prop determining if we should require `currentPassword` 😄

File: apps/web/src/components/forms/password.tsx (hunk 7 above)
Comment (luckyklyist) [reply apps/web/src/components/forms/password.tsx:28]: What Zod validations should I add there to make it work?
```
currentPassword: z
      .string()
      .min(6, { message: 'Password should contain at least 6 characters' })
      .max(72, { message: 'Password should not contain more than 72 characters' })
      .nullable(),
  ```
  This gives error

File: packages/trpc/server/profile-router/router.ts (hunk 6 above)
Comment (luckyklyist) [reply packages/trpc/server/profile-router/router.ts:124]: Do you mean creating the Next.js server component to retrieve data directly?

File: packages/lib/server-only/user/update-password.ts (hunk 2 above)
Comment (luckyklyist) [reply packages/lib/server-only/user/update-password.ts:19]: 😅

**Comment by Mythie** [comment]
Closing for now as we (Documenso) missed the mark on this one, apologies on my end. If this is still an issue with the current application let me know and I'll fix it personally 💚
